omnigibson/object_states/attached_to.py

```diff
@@ -0,0 +1,293 @@
+        return f"{self.parent_link.prim_path}/{self.obj.name}_attachment_joint" if self.parent_link is not None else None
+
+    def _attach(self, other, child_link, parent_link, joint_type=m.DEFAULT_JOINT_TYPE, break_force=m.DEFAULT_BREAK_FORCE, break_torque=m.DEFAULT_BREAK_TORQUE):
+        """
+        Creates a fixed or spherical joint between a male meta link of self.obj (@child_link) and a female meta link of
+         @other (@parent_link) with a given @joint_type, @break_force and @break_torque
+
+         Args:
+            other (DatasetObject): parent object to attach to.
+            child_link (RigidPrim): male meta link of @self.obj.
+            parent_link (RigidPrim): female meta link of @other.
+            joint_type (JointType): joint type of the attachment, {JointType.JOINT_FIXED, JointType.JOINT_SPHERICAL}
+            break_force (float or None): break force for linear dofs, unit is Newton.
+            break_torque (float or None): break torque for angular dofs, unit is Newton-meter.
+        """
+        assert joint_type in {JointType.JOINT_FIXED, JointType.JOINT_SPHERICAL}, f"Unsupported joint type {joint_type}"
+        # Set the parent references
+        self.parent = other
+        self.parent_link = parent_link
+
+        # Set the child reference for @other
+        other.states[AttachedTo].children[parent_link.body_name] = self.obj
+
+        # Set pose for self.obj so that child_link and parent_link align (6dof alignment for FixedJoint and 3dof alignment for SphericalJoint)
+        parent_pos, parent_quat = parent_link.get_position_orientation()
+        child_pos, child_quat = child_link.get_position_orientation()
+
+        child_root_pos, child_root_quat = self.obj.get_position_orientation()
+
+        if joint_type == JointType.JOINT_FIXED:
+            # For FixedJoint: find the relation transformation of the two frames and apply it to self.obj.
+            rel_pos, rel_quat = T.mat2pose(T.pose2mat((parent_pos, parent_quat)) @ T.pose_inv(T.pose2mat((child_pos, child_quat))))
+            new_child_root_pos, new_child_root_quat = T.pose_transform(rel_pos, rel_quat, child_root_pos, child_root_quat)
+        else:
+            # For SphericalJoint: move the position of self.obj to align the two frames and keep the rotation unchanged.
+            new_child_root_pos = child_root_pos + (parent_pos - child_pos)
+            new_child_root_quat = child_root_quat
+
+        # Actually move the object and also keep it still for stability purposes.
+        self.obj.set_position_orientation(new_child_root_pos, new_child_root_quat)
+        self.obj.keep_still()
+
+        if joint_type == JointType.JOINT_FIXED:
+            # For FixedJoint: the parent link, the child link and the joint frame align
+            parent_local_quat = np.array([0.0, 0.0, 0.0, 1.0])
+        else:
+            # SphericalJoint: the same except that the rotation of the parent link doesn't align with the joint frame
```

Comment:
so just to clarify, joint frame with always align with child frame irregardless of joint type right?

Reply:
Correct.